# **Women Cloth Reviews Prediction with Multi Nomial Naïve Bayes**



---



## **Import Library**

In [1]:
import numpy as np

In [2]:
import pandas as pd

In [3]:
import matplotlib.pyplot as plt

In [4]:
import seaborn as sns

## **Import DataSet**

In [5]:
df=pd.read_csv('https://raw.githubusercontent.com/YBI-Foundation/Teaching-Data/main/Women%20Clothing%20E-Commerce%20Review.csv')

In [6]:
df.head()

Clothing ID  Age                    Title  \
0          767   33                      NaN   
1         1080   34                      NaN   
2         1077   60  Some major design flaws   
3         1049   50         My favorite buy!   
4          847   47         Flattering shirt   

                                              Review  Rating  Recommended  \
0  Absolutely wonderful - silky and sexy and comf...       4            1   
1  Love this dress!  it's sooo pretty.  i happene...       5            1   
2  I had such high hopes for this dress and reall...       3            0   
3  I love, love, love this jumpsuit. it's fun, fl...       5            1   
4  This shirt is very flattering to all due to th...       5            1   

   Positive Feedback        Division Department   Category  
0                  0       Initmates   Intimate  Intimates  
1                  4         General    Dresses    Dresses  
2                  0         General    Dresses    Dresses  
3                  0  General Petite    Bottoms      Pants  
4                  6         General       Tops    Blouses

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23486 entries, 0 to 23485
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Clothing ID        23486 non-null  int64 
 1   Age                23486 non-null  int64 
 2   Title              19676 non-null  object
 3   Review             22641 non-null  object
 4   Rating             23486 non-null  int64 
 5   Recommended        23486 non-null  int64 
 6   Positive Feedback  23486 non-null  int64 
 7   Division           23472 non-null  object
 8   Department         23472 non-null  object
 9   Category           23472 non-null  object
dtypes: int64(5), object(5)
memory usage: 1.8+ MB


In [8]:
df.shape

(23486, 10)

## **Missing Values**

Remove missing values in Reviews Columns with no Review Text

In [9]:
df.isna().sum()

Clothing ID             0
Age                     0
Title                3810
Review                845
Rating                  0
Recommended             0
Positive Feedback       0
Division               14
Department             14
Category               14
dtype: int64

In [10]:
df[df['Review']==""]=np.NaN

In [11]:
df['Review'].fillna("No Review",inplace=True)

In [12]:
df.isna().sum()

Clothing ID             0
Age                     0
Title                3810
Review                  0
Rating                  0
Recommended             0
Positive Feedback       0
Division               14
Department             14
Category               14
dtype: int64

In [13]:
df['Review']

0        Absolutely wonderful - silky and sexy and comf...
1        Love this dress!  it's sooo pretty.  i happene...
2        I had such high hopes for this dress and reall...
3        I love, love, love this jumpsuit. it's fun, fl...
4        This shirt is very flattering to all due to th...
                               ...                        
23481    I was very happy to snag this dress at such a ...
23482    It reminds me of maternity clothes. soft, stre...
23483    This fit well, but the top was very see throug...
23484    I bought this dress for a wedding i have this ...
23485    This dress in a lovely platinum is feminine an...
Name: Review, Length: 23486, dtype: object

## **Define Target (y) and Feature (X)**

In [14]:
df.columns

Index(['Clothing ID', 'Age', 'Title', 'Review', 'Rating', 'Recommended',
       'Positive Feedback', 'Division', 'Department', 'Category'],
      dtype='object')

In [15]:
X=df['Review']

In [16]:
y=df['Rating']

In [17]:
df['Rating'].value_counts()

Rating
5    13131
4     5077
3     2871
2     1565
1      842
Name: count, dtype: int64

## **Train Test Split**

In [18]:
from sklearn.model_selection import train_test_split

In [19]:
X_train, X_test, y_train, y_test = train_test_split(X,y, train_size = 0.7, stratify = y, random_state=42529)

In [20]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((16440,), (7046,), (16440,), (7046,))

## **Get Feature Text Coversion to Tokens**

In [21]:
from sklearn.feature_extraction.text import CountVectorizer

In [22]:
cv=CountVectorizer(lowercase=True, analyzer='word',ngram_range=(2,3),stop_words= 'english',max_features=5000)

In [23]:
X_train = cv.fit_transform(X_train)

In [24]:
cv.get_feature_names_out()

array(['00 petite', '10 12', '10 bought', ..., 'years come', 'years old',
       'yoga pants'], dtype=object)

In [25]:
X_train.toarray()

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

In [26]:
X_test=cv.fit_transform(X_test)

In [27]:
cv.get_feature_names_out()

array(['10 12', '10 dress', '10 dresses', ..., 'years old',
       'yellow color', 'yoga pants'], dtype=object)

In [28]:
X_test.toarray()

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

## **Get Model Train**

In [29]:
from sklearn.naive_bayes import MultinomialNB

In [30]:
model = MultinomialNB()

In [31]:
model.fit(X_train, y_train)

MultinomialNB()

In [32]:
MultinomialNB()

MultinomialNB()

## **Get Model Prediction**

In [33]:
y_pred = model.predict(X_test)

In [34]:
y_pred.shape

(7046,)

In [35]:
y_pred

array([4, 5, 4, ..., 3, 4, 1])

## **Get Probability of Each Predicted Class**

In [36]:
model.predict_proba(X_test)

array([[0.01597287, 0.06247668, 0.02343906, 0.86805191, 0.03005948],
       [0.01938616, 0.03618749, 0.01366139, 0.02494917, 0.9058158 ],
       [0.12714164, 0.31224624, 0.02402702, 0.4310927 , 0.10549239],
       ...,
       [0.12935707, 0.07677888, 0.54537581, 0.10858167, 0.13990657],
       [0.01039322, 0.00244866, 0.02592397, 0.69603597, 0.26519819],
       [0.3715833 , 0.28173991, 0.19139055, 0.14724676, 0.00803948]])

## **Get Model Evaluation**

In [37]:
from sklearn.metrics import confusion_matrix, classification_report

In [38]:
print(confusion_matrix(y_test,y_pred))

[[  17   24   29   47  136]
 [  46   60   52   92  220]
 [  91   87  115  184  384]
 [ 156  153  193  334  687]
 [ 395  302  375  737 2130]]


In [39]:
print(classification_report(y_test,y_pred))


              precision    recall  f1-score   support

           1       0.02      0.07      0.04       253
           2       0.10      0.13      0.11       470
           3       0.15      0.13      0.14       861
           4       0.24      0.22      0.23      1523
           5       0.60      0.54      0.57      3939

    accuracy                           0.38      7046
   macro avg       0.22      0.22      0.22      7046
weighted avg       0.41      0.38      0.39      7046



## **Recategories Ratings as poor(0) and good(1)**

In [40]:
df['Rating'].value_counts()


Rating
5    13131
4     5077
3     2871
2     1565
1      842
Name: count, dtype: int64

In [41]:
df.replace({'Rating':{1:0,2:0,3:0,4:1,5:1}}, inplace=True)


In [42]:
y=df['Rating']


In [43]:
X=df['Review']


## **Train Test Split**

In [44]:
from sklearn.model_selection import train_test_split


In [45]:
X_train, X_test, y_train, y_test = train_test_split(X,y, train_size = 0.7, stratify = y, random_state=42529)


In [46]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape


((16440,), (7046,), (16440,), (7046,))

## **Get Feature Text Conversion to Tokens**

In [47]:
from sklearn.feature_extraction.text import CountVectorizer
cv=CountVectorizer(lowercase=True, analyzer='word',ngram_range=(2, 3),stop_words= 'english',
                   max_features=5000)


In [48]:
X_train = cv.fit_transform(X_train)
X_test=cv.fit_transform(X_test)

## **Get Model Re-Train**

In [49]:
from sklearn.naive_bayes import MultinomialNB
model = MultinomialNB()
model.fit(X_train, y_train)

MultinomialNB()

In [50]:
MultinomialNB()


MultinomialNB()

## **Get Model Prediction**

In [51]:
y_pred = model.predict(X_test)
y_pred.shape

(7046,)

In [52]:
y_pred

array([1, 1, 1, ..., 1, 1, 0])

## **Get Model Evaluation**

In [53]:
from sklearn.metrics import confusion_matrix, classification_report


In [54]:
print(confusion_matrix(y_test,y_pred))

[[ 443 1140]
 [1202 4261]]


In [55]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.27      0.28      0.27      1583
           1       0.79      0.78      0.78      5463

    accuracy                           0.67      7046
   macro avg       0.53      0.53      0.53      7046
weighted avg       0.67      0.67      0.67      7046



## **Predictions and Explanations:**
Confusion Matrix: Shows the count of true and false positives and negatives. It indicates where the model is making correct and incorrect predictions.<br>
Classification Report: Provides precision, recall, and f1-score for each class. These metrics help understand the balance between precision and recall for each class.


## **Results**

Initial Multiclass Classification Accuracy: 38%
<br>Binary Classification (Poor/Good): Accuracy 67%
